app/Main.hs

```diff
@@ -96,49 +96,95 @@ writePackageFile =
   . packageConfigToJSON
 
 data PackageInfo = PackageInfo
-  { repo         :: Text
+  { repo         :: Repo
   , version      :: Text
   , dependencies :: [PackageName]
   } deriving (Show, Eq, Generic, Aeson.FromJSON, Aeson.ToJSON)
 
 type PackageSet = Map.Map PackageName PackageInfo
 
-cloneShallow
+newtype Repo = Repo { unRepo :: Text } deriving (Show, Eq)
+
+instance Aeson.FromJSON Repo where
+  parseJSON = fmap Repo . Aeson.parseJSON
+
+
+instance Aeson.ToJSON Repo where
+  toJSON = Aeson.toJSON . unRepo
+
+
+data CloneTarget = CloneTag Text
+                 | CloneSHA Text
+                 deriving (Show)
+
+
+parseCloneTarget
   :: Text
+  -> Either Text CloneTarget
+parseCloneTarget t =
+  if T.null remainder
+     then Right (CloneTag t)
+     else case T.toLower schemeName of
+      "sha" -> Right (CloneSHA withoutScheme)
+      "tag" -> Right (CloneTag withoutScheme)
+      _ -> Left ("Invalid scheme. Expected sha: | tag: but got " <> schemeName)
+  where
+    (schemeName, remainder) = T.breakOn ":" t
+    withoutScheme = T.drop 3 remainder
```

Comment:
I think we probably want to use `stripPrefix "://"` here instead? I could see some very hard to debug behaviour resulting from malformed URLs otherwise.

https://www.stackage.org/haddock/lts-10.5/text-1.2.2.2/Data-Text.html#v:stripPrefix

Reply:
@kRITZCREEK  `stripPrefix` won't work because there's content before the "://", we're trying to extract the scheme beforehand. If we want to be really paranoid about parsing the URI, we could pull in `uri-bytestring`, but that is probably overkill.

@hdgarrood I think you're right. Probably should have console tested this since we don't really have a formal test suite for this project.